tuf/tuf.go

```diff
@@ -296,114 +302,102 @@ func (tr *Repo) GetAllLoadedRoles() []*data.Role {
 	return res
 }
 
-// GetDelegation finds the role entry representing the provided
-// role name along with its associated public keys, or ErrInvalidRole
-func (tr *Repo) GetDelegation(role string) (*data.Role, data.Keys, error) {
-	if !data.IsDelegation(role) {
-		return nil, nil, data.ErrInvalidRole{Role: role, Reason: "not a valid delegated role"}
-	}
-
-	parent := path.Dir(role)
-
-	// check the parent role
-	if _, err := tr.GetDelegationRole(parent); parent != data.CanonicalTargetsRole && err != nil {
-		return nil, nil, data.ErrInvalidRole{Role: role, Reason: "parent role not found"}
-	}
-
-	// check the parent role's metadata
-	p, ok := tr.Targets[parent]
-	if !ok { // the parent targetfile may not exist yet, so it can't be in the list
-		return nil, nil, data.ErrNoSuchRole{Role: role}
-	}
-
-	foundAt := utils.FindRoleIndex(p.Signed.Delegations.Roles, role)
-	if foundAt < 0 {
-		return nil, nil, data.ErrNoSuchRole{Role: role}
-	}
-	delegationRole := p.Signed.Delegations.Roles[foundAt]
-	keys := make(data.Keys)
-	for _, keyID := range delegationRole.KeyIDs {
-		keys[keyID] = p.Signed.Delegations.Keys[keyID]
-	}
-	return delegationRole, keys, nil
-}
-
 // UpdateDelegations updates the appropriate delegations, either adding
 // a new delegation or updating an existing one. If keys are
 // provided, the IDs will be added to the role (if they do not exist
 // there already), and the keys will be added to the targets file.
-func (tr *Repo) UpdateDelegations(role *data.Role, keys []data.PublicKey) error {
-	if !data.IsDelegation(role.Name) {
-		return data.ErrInvalidRole{Role: role.Name, Reason: "not a valid delegated role"}
+func (tr *Repo) UpdateDelegations(roleName string, changes changelist.TufDelegation) error {
+	if !data.IsDelegation(roleName) {
+		return data.ErrInvalidRole{Role: roleName, Reason: "not a valid delegated role"}
 	}
-	parent := path.Dir(role.Name)
+	parent := path.Dir(roleName)
 
 	if err := tr.VerifyCanSign(parent); err != nil {
 		return err
 	}
 
 	// check the parent role's metadata
-	p, ok := tr.Targets[parent]
+	_, ok := tr.Targets[parent]
 	if !ok { // the parent targetfile may not exist yet - if not, then create it
 		var err error
-		p, err = tr.InitTargets(parent)
+		_, err = tr.InitTargets(parent)
 		if err != nil {
 			return err
 		}
 	}
 
-	for _, k := range keys {
-		if !utils.StrSliceContains(role.KeyIDs, k.ID()) {
-			role.KeyIDs = append(role.KeyIDs, k.ID())
+	// Walk to parent, and either set or update this delegation
+	// Ensure all updates are valid
+	delegationUpdateVisitor := func(tgt *data.SignedTargets, validRole data.DelegationRole) error {
+		var err error
+		var createdRole bool
+		// Validate the changes underneath this restricted validRole for adding paths
+		changes.AddPaths = data.RestrictDelegationPathPrefixes(validRole.Paths, changes.AddPaths)
```

Comment:
Non-blocking question: do we want to actually restrict the paths when creating the delegation?  If a user is requesting a paths change, this may be a surprise when the paths they specify are changed.  Would we want to just reject the change instead?